### PR TITLE
Feature/actp 394/timer adjustment by proctor

### PR DESCRIPTION
**Pull request by augustas**
Time adjustment endpoint for proctor.
 
Related to : https://oat-sa.atlassian.net/browse/ACTP-394
 
New endpoint in monitoring, to store forward the timer adjustment.
  
#### Dependencies
 
Requires :
 - [x] https://github.com/oat-sa/extension-tao-test/pull/321
 - [x] https://github.com/oat-sa/extension-tao-testqti/pull/1726

### CODE REVIEW

File: model/execution/DeliveryExecutionManagerService.php [hunk 1]
```diff
@@ -238,4 +236,60 @@ public function setExtraTime($deliveryExecutions, $extraTime = 0, $extendedTime
...
+                    $success = $timerAdjustmentService->increase($deliveryExecution, $seconds);
+                } else {
+                    $success = $timerAdjustmentService->decrease($deliveryExecution, abs($seconds));
+                }
+            }
+
+            if ($success) {
+                $result['processed'][$deliveryExecution->getIdentifier()] = true;
+            } else {
+                $result['unprocessed'][$deliveryExecution->getIdentifier()] = false;
```
Comment (zagovorichev) [inline model/execution/DeliveryExecutionManagerService.php:264]: why uprocessed === false is true? why unprocessed = true is not unprocessed?

File: model/implementation/DeliveryExecutionStateService.php [hunk 2]
```diff
@@ -57,6 +57,9 @@ class DeliveryExecutionStateService extends AbstractStateService implements \oat
     const OPTION_CANCELLATION_DELAY = 'cancellation_delay';
     const OPTION_TIME_HANDLING = 'time_handling';
 
+    const TIME_HANDLING_EXTRA_TIME = 'extra_time';
```
Comment (zagovorichev) [inline model/implementation/DeliveryExecutionStateService.php:60]: I'm just wondering : what if we rename this variable https://github.com/oat-sa/extension-tao-proctoring/blob/8b98ae91a1ce06954d9456ad34ad1cabaee3784a/model/monitorCache/DeliveryMonitoringService.php#L55, then this one should be renamed as well? then I prefer only 1 constant

File: model/execution/DeliveryExecutionManagerService.php [hunk 3]
```diff
@@ -238,4 +236,60 @@ public function setExtraTime($deliveryExecutions, $extraTime = 0, $extendedTime
...
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param DeliveryExecutionInterface $deliveryExecution
+     * @return bool
+     */
+    private function isCorrectDeliveryExecutionStateForAdjustment($deliveryExecution)
```
Comment (SashaTAO) [inline model/execution/DeliveryExecutionManagerService.php:275]: Maybe we could rename it to something shorter like `canAdjustTimer` or `timerAdjustmentAllowed`. We may have some restrictions not related to state later.

File: model/implementation/DeliveryExecutionStateService.php [hunk 4]
```diff
@@ -57,6 +57,9 @@ class DeliveryExecutionStateService extends AbstractStateService implements \oat
...
+    const TIME_HANDLING_TIMER_ADJUSTMENT = 'timer_adjustment';
```
Comment (SashaTAO) [inline model/implementation/DeliveryExecutionStateService.php:61]: Why do we need these constants if they are not used?

File: model/execution/DeliveryExecutionManagerService.php (hunk 1 above)
Comment (augustas) [reply model/execution/DeliveryExecutionManagerService.php:264]: Current frontend implementation expects such values, same as in `DeliveryExecutionManagerService::setExtraTime()`. I agree it's not very clear, maybe original developer though `unprocessed` is a failure and should get a `false`.

File: model/implementation/DeliveryExecutionStateService.php (hunk 2 above)
Comment (augustas) [reply model/implementation/DeliveryExecutionStateService.php:60]: The constant you've linked is concerning a key to use when storing data in `delivery_monitoring` table. While the new constant in this file is for the values valid for option `OPTION_TIME_HANDLING`

File: model/execution/DeliveryExecutionManagerService.php (hunk 3 above)
Comment (augustas) [reply model/execution/DeliveryExecutionManagerService.php]: Done

File: model/implementation/DeliveryExecutionStateService.php (hunk 4 above)
Comment (augustas) [reply model/implementation/DeliveryExecutionStateService.php:61]: I've updated the updater to use the constant `TIME_HANDLING_EXTRA_TIME`, and client extensions which require the timer adjustment feature will use the second constant `TIME_HANDLING_TIMER_ADJUSTMENT`.

**Comment by zagovorichev** [comment]
@augustas could you add instruction how can we test this task (not the unit tests only), according to our github workflow: https://oat-sa.atlassian.net/wiki/spaces/DEV/pages/11338720/Development+Workflow+with+Github

Thank you in advance :)

**Comment by augustas** [comment]
> @augustas could you add instruction how can we test this task (not the unit tests only), according to our github workflow: https://oat-sa.atlassian.net/wiki/spaces/DEV/pages/11338720/Development+Workflow+with+Github
> 
> Thank you in advance :)

@zagovorichev The frontend changes that would enable the testing has not been done yet: https://oat-sa.atlassian.net/browse/TCA-395
In order to test it manually, you could enable the extra time functionality (https://github.com/oat-sa/extension-tao-proctoring/blob/master/model/implementation/DeliveryExecutionStateService.php#L58) and manually change the endpoint address here: https://github.com/oat-sa/extension-tao-proctoring/blob/master/views/js/controller/Delivery/monitoring.js#L92 to point to the new endpoint `adjustTime`. Then adding extra time would register an increase, yet without debugging you won't see the change, since the stored timer adjustments do not change the timer behavior yet.

File: model/execution/DeliveryExecutionList.php [hunk 5]
```diff
@@ -102,6 +102,9 @@ private function createExecution($cachedData, $extraFields)
             'allowExtraTime' => isset($cachedData[DeliveryMonitoringService::ALLOW_EXTRA_TIME])
                 ? (bool)$cachedData[DeliveryMonitoringService::ALLOW_EXTRA_TIME]
                 : null,
+            'allowTimerAdjustment' => $this->getDeliveryExecutionManagerService()->isTimerAdjustmentAllowed(
```
Comment (SashaTAO) [inline model/execution/DeliveryExecutionList.php:105]: With this change many tests in `DeliveryExecutionListTest` fail. We need to mock `DeliveryExecutionManagerService` service in ServiceLocator mock there.

Comment (augustas) [reply model/execution/DeliveryExecutionList.php:105]: I've fixed the `DeliveryExecutionListTest` unit test.

File: model/execution/DeliveryExecutionManagerService.php [hunk 6]
```diff
@@ -238,4 +235,60 @@ public function setExtraTime($deliveryExecutions, $extraTime = 0, $extendedTime
...
+            if (is_string($deliveryExecution)) {
+                $deliveryExecution = $this->getDeliveryExecutionById($deliveryExecution);
+            }
+
+            $success = false;
+            if ($this->isTimerAdjustmentAllowed($deliveryExecution)) {
+                if ($seconds > 0) {
+                    $success = $timerAdjustmentService->increase($deliveryExecution, $seconds);
+                } else {
+                    $success = $timerAdjustmentService->decrease($deliveryExecution, abs($seconds));
```
[the remark this answers is not in the document]
Comment (augustas) [reply model/execution/DeliveryExecutionManagerService.php]: It's a problem with Github caching, this was updated on friday

**Pull request merged**